nmagent/client.go

```diff
@@ -0,0 +1,208 @@
+	req, err := c.buildRequest(ctx, jnr)
+	if err != nil {
+		return errors.Wrap(err, "building request")
+	}
+
+	err = c.retrier.Do(ctx, func() error {
+		resp, err := c.httpClient.Do(req) // nolint:govet // the shadow is intentional
+		if err != nil {
+			return errors.Wrap(err, "executing request")
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			return c.error(resp.StatusCode, resp.Header, resp.Body)
+		}
+		return nil
+	})
+
+	return err // nolint:wrapcheck // wrapping this just introduces noise
+}
+
+// GetNetworkConfiguration retrieves the configuration of a customer's virtual
+// network. Only subnets which have been delegated will be returned
+func (c *Client) GetNetworkConfiguration(ctx context.Context, gncr GetNetworkConfigRequest) (VirtualNetwork, error) {
+	var out VirtualNetwork
+
+	req, err := c.buildRequest(ctx, gncr)
+	if err != nil {
+		return out, errors.Wrap(err, "building request")
+	}
+
+	err = c.retrier.Do(ctx, func() error {
+		resp, err := c.httpClient.Do(req) // nolint:govet // the shadow is intentional
+		if err != nil {
+			return errors.Wrap(err, "executing http request to")
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			return c.error(resp.StatusCode, resp.Header, resp.Body)
+		}
+
+		ct := resp.Header.Get(internal.HeaderContentType)
+		if ct != internal.MimeJSON {
+			return NewContentError(ct, resp.Body, resp.ContentLength)
+		}
+
+		err = json.NewDecoder(resp.Body).Decode(&out)
+		if err != nil {
+			return errors.Wrap(err, "decoding json response")
+		}
+
+		return nil
+	})
+
+	return out, err // nolint:wrapcheck // wrapping just introduces noise here
+}
+
+// PutNetworkContainer applies a Network Container goal state and publishes it
+// to PubSub
+func (c *Client) PutNetworkContainer(ctx context.Context, pncr *PutNetworkContainerRequest) error {
+	req, err := c.buildRequest(ctx, pncr)
+	if err != nil {
+		return errors.Wrap(err, "building request")
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "submitting request")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return c.error(resp.StatusCode, resp.Header, resp.Body)
+	}
+	return nil
+}
+
+// DeleteNetworkContainer removes a Network Container, its associated IP
+// addresses, and network policies from an interface
+func (c *Client) DeleteNetworkContainer(ctx context.Context, dcr DeleteContainerRequest) error {
+	req, err := c.buildRequest(ctx, dcr)
+	if err != nil {
+		return errors.Wrap(err, "building request")
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "submitting request")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return c.error(resp.StatusCode, resp.Header, resp.Body)
+	}
+
+	return nil
+}
+
+func (c *Client) error(code int, headers http.Header, body io.ReadCloser) error {
+	// read the entire body
+	defer body.Close()
```

Comment:
- Does this need to be a method? Doesn't seem like the receiver is used.
- Seems like this close is redundant with all usages of `error` (caller always closes). It could be removed, and this just accept a `Reader`